crates/bevy_reflect/src/lib.rs

```diff
@@ -1,4 +1,405 @@
+//! or [`Reflect::into_reflect`] methods.
+//!
+//! ## Value Types
+//!
+//! Types that do not fall under one of the above subtraits,
+//! such as for primitives (e.g. `bool`, `usize`, etc.)
+//! and simple types (e.g. `String`, `Duration`),
+//! are referred to as _value_ types
+//! since methods like [`Reflect::reflect_ref`] return a [`ReflectRef::Value`] variant.
+//! While most other types contain their own `dyn Reflect` fields and data,
+//! these types generally cannot be broken down any further.
+//!
+//! # Dynamic Types
+//!
+//! Each subtrait comes with a corresponding _dynamic_ type.
+//!
+//! The available dynamic types are:
+//! * [`DynamicTuple`]
+//! * [`DynamicArray`]
+//! * [`DynamicList`]
+//! * [`DynamicMap`]
+//! * [`DynamicStruct`]
+//! * [`DynamicTupleStruct`]
+//! * [`DynamicEnum`]
+//!
+//! These dynamic types may contain any arbitrary reflected data.
+//!
+//! ```
+//! # use bevy_reflect::{DynamicStruct, Struct};
+//! let mut data = DynamicStruct::default();
+//! data.insert("foo", 123_i32);
+//! assert_eq!(Some(&123), data.field("foo").unwrap().downcast_ref::<i32>())
+//! ```
+//!
+//! They are most commonly used as "proxies" for other types,
+//! where they contain the same data as— and therefore, represent— a concrete type.
+//! The [`Reflect::clone_value`] method will return a dynamic type for all non-value types,
+//! allowing all types to essentially be "cloned".
+//! And since dynamic types themselves implement [`Reflect`],
+//! we may pass them around just like any other reflected type.
+//!
+//! ```
+//! # use bevy_reflect::{DynamicStruct, Reflect};
+//! # #[derive(Reflect)]
+//! # struct MyStruct {
+//! #   foo: i32
+//! # }
+//! let original: Box<dyn Reflect> = Box::new(MyStruct {
+//!   foo: 123
+//! });
+//!
+//! // `cloned` will be a `DynamicStruct` representing a `MyStruct`
+//! let cloned: Box<dyn Reflect> = original.clone_value();
+//! assert!(cloned.represents::<MyStruct>());
+//! assert!(cloned.is::<DynamicStruct>());
+//! ```
+//!
+//! ## Patching
+//!
+//! These dynamic types come in handy when needing to apply multiple changes to another type.
+//! This is known as "patching" and is done using the [`Reflect::apply`] method.
+//!
+//! ```
+//! # use bevy_reflect::{DynamicEnum, Reflect};
+//! let mut value = Some(123_i32);
+//! let patch = DynamicEnum::new(std::any::type_name::<Option<i32>>(), "None", ());
+//! value.apply(&patch);
+//! assert_eq!(None, value);
+//! ```
+//!
+//! ## `FromReflect`
+//!
+//! It's important to remember that dynamic types are _not_ the concrete type they may be representing.
+//! A common mistake is to treat them like such when trying to cast back to the original type
+//! or when trying to make use of a reflected trait which expects the actual type.
+//!
+//! ```should_panic
+//! # use bevy_reflect::{DynamicStruct, Reflect};
+//! # #[derive(Reflect)]
+//! # struct MyStruct {
+//! #   foo: i32
+//! # }
+//! let original: Box<dyn Reflect> = Box::new(MyStruct {
```

Comment:
Why are you wrapping everything in Box::new() for the examples?

Reply:
This is mainly so that we could strongly type everything as `Box<dyn Reflect>`. My hope in doing so was to:
1. Show that this is purely `Reflect` doing these things
2. Show that reflection data is normally passed around as trait objects

If this is too confusing/verbose, though, we can remove it!

Reply:
I think it's fine, I think I understand now. We are doing dynamic dispatch here and I am assuming that we can't do that unless we have allocated a specific amount of memory allocated already.

Reply:
Sorta. The `Box` isn't required here. It's just to be explicit in showing that it's the struct as a `dyn Reflect` object that we care about.